File: ChessLab/Play/EngineStrength.swift
import ChessKitEngine

/// Réglage de la force du moteur pour le mode Jouer.
///
/// Stockfish ne permet de limiter `UCI_Elo` qu'entre 1320 et 3190. En
/// dessous, on simule un niveau plus faible en désactivant `UCI_LimitStrength`
/// et en abaissant `Skill Level` + la profondeur de recherche à la place.
enum EngineStrength: Equatable {
    /// Force plafonnée via `UCI_LimitStrength` + `UCI_Elo` (1320...3190).
    case limited(elo: Int)
    /// Simulation d'un niveau sous 1320 : `Skill Level` bas + profondeur limitée.
    case belowMinimum(approximateElo: Int, skillLevel: Int, depth: Int)
    /// Aucune limite : pleine puissance du moteur.
    case maximum

    /// Plage complète (Laboratoire) : jusqu'à la pleine puissance, qui est
    /// tout l'intérêt de faire s'affronter deux moteurs.
    static let sliderRange: ClosedRange<Double> = 800...3190
    /// Plage du mode **Jouer**, plafonnée à 2500 : au-delà, on ne propose plus
    /// un adversaire, on propose une défaite. Le Laboratoire, lui, garde la
    /// plage complète.
    static let playSliderRange: ClosedRange<Double> = 800...2500
    static let ratedMinimum = 1320

    /// Construit un réglage à partir d'une valeur de slider continue.
    init(sliderValue: Double) {
        let elo = Int(sliderValue.rounded())

        if elo >= 3190 {
            self = .maximum
        } else if elo >= Self.ratedMinimum {
            self = .limited(elo: elo)
        } else {
            // Interpole skill level (0...5) et profondeur (1...6) sur la
            // plage 800...1319 pour approcher un niveau très faible.
            let t = Double(elo - 800) / Double(Self.ratedMinimum - 1 - 800)
            let skill = Int((t * 5).rounded())
            let depth = Int((1 + t * 5).rounded())
            self = .belowMinimum(approximateElo: elo, skillLevel: max(0, skill), depth: max(1, depth))
        }
    }

    /// Valeur à afficher sur le slider pour ce réglage.
    var sliderValue: Double {
        switch self {
        case let .limited(elo): Double(elo)
        case let .belowMinimum(approximateElo, _, _): Double(approximateElo)
        case .maximum: Self.sliderRange.upperBound
        }
    }

    /// Commandes `setoption` UCI à envoyer au moteur pour appliquer ce réglage.
    var setupCommands: [EngineCommand] {
        switch self {
        case let .limited(elo):
            [
                .setoption(id: "UCI_LimitStrength", value: "true"),
                .setoption(id: "UCI_Elo", value: "\(elo)"),
                .setoption(id: "Skill Level", value: "20"),
            ]
        case let .belowMinimum(_, skillLevel, _):
            [
                .setoption(id: "UCI_LimitStrength", value: "false"),
                .setoption(id: "Skill Level", value: "\(skillLevel)"),
            ]
        case .maximum:
            [
                .setoption(id: "UCI_LimitStrength", value: "false"),
                .setoption(id: "Skill Level", value: "20"),
            ]
        }
    }

    /// Profondeur de recherche maximale à utiliser pour `go depth`, si
    /// applicable (les niveaux sous 1320 plafonnent aussi la profondeur).
    var maxDepth: Int? {
        if case let .belowMinimum(_, _, depth) = self {
            depth
        } else {
            nil
        }
    }

    /// Description courte pour affichage.
    var displayLabel: String {
        switch self {
        case let .limited(elo): "Elo \(elo)"
        case let .belowMinimum(approximateElo, _, _): "Elo ~\(approximateElo)"
        case .maximum: "Maximum"
        }
    }
}

/// Préréglages proposés à l'utilisateur en plus du slider libre.
struct EnginePreset: Identifiable, Equatable {
    let id: String
    let label: String
    let strength: EngineStrength

    /// Libellé affiché sur le chip, avec l'Elo approximatif.
    var chipLabel: String {
        strength == .maximum ? label : "\(label) (\(Int(strength.sliderValue)))"
    }

    /// Échelle du mode Jouer. Plus de « Élite mondiale » (2800) ni de
    /// « Maximum » : ces deux-là ne se jouent pas, ils se subissent. Les deux
    /// marches ajoutées en bas (1000 et 1400) resserrent au contraire l'écart
    /// là où l'on progresse vraiment — entre le grand débutant et le joueur de
    /// club, l'ancienne échelle sautait de 800 à 1200 puis à 1600.
    static let all: [EnginePreset] = [
        .init(id: "e800", label: "Grand débutant", strength: EngineStrength(sliderValue: 800)),
        .init(id: "e1000", label: "Débutant", strength: EngineStrength(sliderValue: 1000)),
        .init(id: "e1200", label: "Débutant confirmé", strength: EngineStrength(sliderValue: 1200)),
        .init(id: "e1400", label: "Intermédiaire", strength: .limited(elo: 1400)),
        .init(id: "e1600", label: "Intermédiaire confirmé", strength: .limited(elo: 1600)),
        .init(id: "e2000", label: "Avancé / Expert", strength: .limited(elo: 2000)),
        .init(id: "e2300", label: "Maître national", strength: .limited(elo: 2300)),
        .init(id: "e2500", label: "Grand Maître", strength: .limited(elo: 2500)),
    ]
}

File: android/app/src/main/java/com/chesslab/play/EngineStrength.kt
package com.chesslab.play

import androidx.annotation.StringRes
import com.chesslab.R
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Le réglage de la FORCE du moteur. Pendant de `EngineStrength.swift`.
 *
 * Stockfish ne sait limiter `UCI_Elo` qu'entre 1320 et 3190. En dessous, on
 * simule un niveau plus faible autrement : `UCI_LimitStrength` désactivé, un
 * `Skill Level` bas, et une PROFONDEUR plafonnée — sans quoi un « débutant »
 * à qui l'on donne un dixième de seconde reste un joueur redoutable sur un
 * plateau à trente-deux pièces.
 */
sealed class EngineStrength {

    /** Force plafonnée par `UCI_LimitStrength` + `UCI_Elo` (1320…3190). */
    data class Limited(val elo: Int) : EngineStrength()

    /** Sous 1320 : `Skill Level` bas et profondeur limitée. */
    data class BelowMinimum(val approximateElo: Int, val skillLevel: Int, val depth: Int) : EngineStrength()

    /** Aucune limite : la pleine puissance du moteur. */
    data object Maximum : EngineStrength()

    /** La valeur à montrer sur le curseur pour ce réglage. */
    val sliderValue: Double
        get() = when (this) {
            is Limited -> elo.toDouble()
            is BelowMinimum -> approximateElo.toDouble()
            Maximum -> sliderRange.endInclusive
        }

    /** Les `setoption` à envoyer au moteur pour appliquer ce réglage. */
    val setupCommands: List<String>
        get() = when (this) {
            is Limited -> listOf(
                "setoption name UCI_LimitStrength value true",
                "setoption name UCI_Elo value $elo",
                "setoption name Skill Level value 20",
            )
            is BelowMinimum -> listOf(
                "setoption name UCI_LimitStrength value false",
                "setoption name Skill Level value $skillLevel",
            )
            Maximum -> listOf(
                "setoption name UCI_LimitStrength value false",
                "setoption name Skill Level value 20",
            )
        }

    /**
     * Les mêmes commandes pour Fairy-Stockfish, dont les bornes d'`UCI_Elo`
     * sont 500…2850 et non 1320…3190. Une valeur hors bornes y est REJETÉE EN
     * SILENCE, le bridage restant actif à l'Elo par défaut : un curseur à 3000
     * donnait un adversaire à 1350 sans le dire.
     */
    val fairySetupCommands: List<String>
        get() {
            if (this !is Limited) return setupCommands
            if (elo > fairyRatedRange.last) return Maximum.setupCommands
            return Limited(max(elo, fairyRatedRange.first)).setupCommands
        }

    /** La profondeur maximale de `go depth`, quand il y en a une. */
    val maxDepth: Int?
        get() = (this as? BelowMinimum)?.depth

    /** « Elo 1400 », « Elo ~1000 » sous la borne, ou « Maximum ». */
    @get:StringRes
    val labelRes: Int
        get() = when (this) {
            is Limited -> R.string.strength_elo
            is BelowMinimum -> R.string.strength_elo_approx
            Maximum -> R.string.strength_maximum
        }

    companion object {
        /** La plage complète, jusqu'à la pleine puissance. */
        val sliderRange = 800.0..3190.0

        /**
         * La plage du mode Jouer : la MÊME. On laisse monter aussi haut qu'on
         * veut, quitte à se faire écraser.
         */
        val playSliderRange = 800.0..3190.0

        const val ratedMinimum = 1320

        val fairyRatedRange = 500..2850

        /** Le réglage correspondant à une position de curseur. */
        fun of(sliderValue: Double): EngineStrength {
            val elo = sliderValue.roundToInt()
            return when {
                elo >= 3190 -> Maximum
                elo >= ratedMinimum -> Limited(elo)
                else -> {
                    // Skill Level (0…5) et profondeur (1…6) interpolés sur
                    // 800…1319 pour approcher un niveau très faible.
                    val t = (elo - 800).toDouble() / (ratedMinimum - 1 - 800).toDouble()
                    BelowMinimum(
                        approximateElo = elo,
                        skillLevel = max(0, (t * 5).roundToInt()),
                        depth = max(1, (1 + t * 5).roundToInt()),
                    )
                }
            }
        }
    }
}

/**
 * Les paliers NOMMÉS du curseur. Ils s'arrêtent à « Grand Maître » : au-delà,
 * le curseur monte jusqu'au maximum, mais sans nom — ces niveaux ne se jouent
 * pas, ils se subissent. Les marches du bas resserrent l'écart là où l'on
 * progresse vraiment.
 */
data class EnginePreset(val id: String, @StringRes val label: Int, val strength: EngineStrength) {
    companion object {
        val all = listOf(
            EnginePreset("e800", R.string.preset_absolute_beginner, EngineStrength.of(800.0)),
            EnginePreset("e1000", R.string.preset_beginner, EngineStrength.of(1000.0)),
            EnginePreset("e1200", R.string.preset_improving, EngineStrength.of(1200.0)),
            EnginePreset("e1400", R.string.preset_intermediate, EngineStrength.Limited(1400)),
            EnginePreset("e1600", R.string.preset_strong_intermediate, EngineStrength.Limited(1600)),
            EnginePreset("e2000", R.string.preset_advanced, EngineStrength.Limited(2000)),
            EnginePreset("e2300", R.string.preset_national_master, EngineStrength.Limited(2300)),
            EnginePreset("e2500", R.string.preset_grandmaster, EngineStrength.Limited(2500)),
        )

        /**
         * Le palier dont le nom décrit une valeur du curseur : le plus proche,
         * l'égalité allant au plus bas (900 est encore « Grand débutant », 950
         * déjà « Débutant »). Au maximum, aucun palier — c'est « Maximum », et
         * le libellé du chiffre le dit.
         */
        fun nearest(sliderValue: Double): EnginePreset? {
            if (sliderValue >= EngineStrength.playSliderRange.endInclusive) return null
            return all.minByOrNull { abs(it.strength.sliderValue - sliderValue) }
        }
    }
}
